creusot/tests/should_succeed/red_black_tree.rs

```diff
@@ -178,6 +178,49 @@ impl<K: DeepModel, V> ShallowModel for Tree<K, V> {
     }
 }
 
+impl<K: DeepModel, V> Resolve for Tree<K, V> {
+    #[open(self)]
+    #[predicate(prophetic)]
+    fn resolve(self) -> bool {
+        pearlite! {
+            forall<k: K, v: V> self.has_mapping(k.deep_model(), v) ==> resolve(&v)
+        }
+    }
+
+    #[open(self)]
+    #[logic(prophetic)]
+    #[requires(structural_resolve(&self))]
+    #[ensures(self.resolve())]
+    fn resolve_coherence(self)
+    where
+        Self: Sized,
+    {
+    }
+}
+
+impl<K: DeepModel, V> Resolve for Node<K, V> {
```

Comment:
Ideally, we would also have an instance for the only public type `Map<K, V>` at the end of the file, and it should be based on the shallow model (`has_mapping` is a private symbol).

Reply:
I just took a look at this, there is a complication introduced by the lack of invariants in the logic. We would need to add `inv` as a precondition to resolve, which... seems heavy weight and annoying. Additionally I worry that we're heading to a situation in which we are back to manually adding `inv` to every trait declaration :/

Reply:
You don't need to add `inv` as a precondition of `resolve`, you can simply guard `resolve` by it.

I.e., `resolve(map) = inv(map) -> forall k, resolve(map@(k))`

Reply:
Actually, you don't even need this here, because there is only one implication of `has_mapping_model` that actually needs the invariants, and we need the other implication to prove coherence!

Reply:
Thinking about it, I think it would not be too bad to add `inv` as a precondition:
- It's a preconditon of `resolve_coherence`, not of `resolve`
- It is not so annoying for the user, because when instantiating, we don't need to add the preconditon if we don't need it.
- At some point, we will change the handling of default implementations so that users will not have to repeat the full specification when instantiating.

Reply:
> It's a preconditon of resolve_coherence, not of resolve

Yea that's actually a good point which alleviates things a good amount imo.

Reply:
Done.